dora/core/common/src/main/java/alluxio/membership/ServiceDiscoveryRecipe.java

```diff
@@ -0,0 +1,324 @@
+ */
+
+package alluxio.membership;
+
+import alluxio.Constants;
+import alluxio.annotation.SuppressFBWarnings;
+import alluxio.exception.status.AlreadyExistsException;
+import alluxio.resource.LockResource;
+import alluxio.util.ThreadFactoryUtils;
+
+import com.google.common.base.Preconditions;
+import io.etcd.jetcd.ByteSequence;
+import io.etcd.jetcd.KeyValue;
+import io.etcd.jetcd.Txn;
+import io.etcd.jetcd.kv.TxnResponse;
+import io.etcd.jetcd.lease.LeaseKeepAliveResponse;
+import io.etcd.jetcd.op.Cmp;
+import io.etcd.jetcd.op.CmpTarget;
+import io.etcd.jetcd.op.Op;
+import io.etcd.jetcd.options.GetOption;
+import io.etcd.jetcd.options.PutOption;
+import io.etcd.jetcd.support.CloseableClient;
+import io.grpc.stub.StreamObserver;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.NoSuchElementException;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.stream.Collectors;
+import javax.annotation.concurrent.GuardedBy;
+
+/**
+ * ServiceDiscoveryRecipe for etcd, to track health status
+ * of all registered services.
+ */
+public class ServiceDiscoveryRecipe {
+  private static final Logger LOG = LoggerFactory.getLogger(AlluxioEtcdClient.class);
+  private static final String BASE_PATH = "/ServiceDiscovery";
+  AlluxioEtcdClient mAlluxioEtcdClient;
+  ScheduledExecutorService mExecutor;
+  String mClusterIdentifier = "";
+  @SuppressFBWarnings({"URF_UNREAD_FIELD"})
+  private final ReentrantLock mRegisterLock = new ReentrantLock();
+  final ConcurrentHashMap<String, ServiceEntity> mRegisteredServices = new ConcurrentHashMap<>();
+
+  /**
+   * CTOR for ServiceDiscoveryRecipe.
+   * @param client
+   * @param clusterIdentifier
+   */
+  public ServiceDiscoveryRecipe(AlluxioEtcdClient client, String clusterIdentifier) {
+    mAlluxioEtcdClient = client;
+    mAlluxioEtcdClient.connect();
+    mClusterIdentifier = clusterIdentifier;
+    mExecutor = Executors.newSingleThreadScheduledExecutor(
+        ThreadFactoryUtils.build("service-discovery-checker", false));
+    mExecutor.scheduleWithFixedDelay(this::checkAllForReconnect,
+        AlluxioEtcdClient.DEFAULT_LEASE_TTL_IN_SEC, AlluxioEtcdClient.DEFAULT_LEASE_TTL_IN_SEC,
+        TimeUnit.SECONDS);
+  }
+
+  /**
+   * Get register path prefix.
+   * @return register path prefix
+   */
+  private String getRegisterPathPrefix() {
+    return String.format("%s/%s", BASE_PATH, mClusterIdentifier);
```

Comment:
use separator constant here?
also i recall from somewhere that `String.format` should be avoided in java due to performance reasons. maybe @apc999 or @jiacheliu3 can provide the context and guidance for it?

Reply:
ok  will put in SEPARATOR here, and let me check with them on the format thing

Reply:
think its other way round to use format and avoiding concatenation:
https://docs.google.com/document/d/1z88jjsieWDo5umr-TxSvAwtjZy58IiSbuSZSrCECWJA/edit

I'm changing other places where uses SEPARATOR for concatenation with string format

Reply:
String concat is definitely the worse, but string formatting is distinct from `String.format`. as described in the link, StringBuilder and StringBuffer are more optimal.

some digging through top few google search results:
- https://cowtowncoder.medium.com/measuring-performance-of-java-string-format-or-lack-thereof-2e1c6a13362c
  - each String.format creates a new object per invocation, which is additional overhead
  - only use for convenience in codepaths with limited exposure (ex. initialization, exception handling)
- https://kylewbanks.com/blog/java-string-concatenation-vs-stringbuilder-vs-string-format-performance
  - same conclusions
- https://www.javaspecialists.eu/archive/Issue294-String.format-3x-faster-in-Java-17.html
  - ... but we don't use java17 lol

i leave it up to you to evaluate how often it is used. seeing `String.format` is just one of my natural red flag triggers when i (occasionally) read java code.

Reply:
I've changed String.format to be only used in initialize or other one-time use case only. and used StringBuffer else where to replace all concatenation.